### PR TITLE
API 7560: allow grant-length to be a Period

### PR DESCRIPTION
**Pull request by worthydolt**

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala [hunk 1]
```diff
@@ -16,45 +16,112 @@
...
-  case object TEN_YEARS         extends GrantLength { val days = 3650  }
-  case object ONE_HUNDRED_YEARS extends GrantLength { val days = 36500 }
-
-  implicit val ordering: Ordering[GrantLength] = Ordering.by(_.days)
-  val values                                   = SortedSet[GrantLength](ONE_DAY, ONE_MONTH, THREE_MONTHS, SIX_MONTHS, ONE_YEAR, EIGHTEEN_MONTHS, THREE_YEARS, FIVE_YEARS, TEN_YEARS, ONE_HUNDRED_YEARS)
-
-  private val allowedIntegerValues = values.map(_.days)
+
+  val lengthRead: Reads[Long] = (JsPath \ "length").read[Long]
+  val unitRead: Reads[String] = (JsPath \ "unit").read[String].map(u => u.toUpperCase)
```
Comment (johnsgp) [inline src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala:39]: Could these be private?

File: src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala [hunk 2]
```diff
@@ -16,45 +16,112 @@
...
+
+  implicit val durationReads: Reads[Duration] = (
+    lengthRead and unitRead.map(ChronoUnit.valueOf(_))
+  )(Duration.of(_, _))
+
+  implicit val durationWrites: Writes[Duration] = (
+    (JsPath \ "length").write[Long] and
```
Comment (johnsgp) [inline src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala:46]: The `java.time.Duration.of(long amount, TemporalUnit unit)` method calls this `amount` (see [Javadoc](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/time/Duration.html#of(long,java.time.temporal.TemporalUnit))). Should we keep it the same? The word `length` has a different scientific meaning.

File: src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala [hunk 3]
```diff
@@ -43,6 +43,12 @@ class GrantLengthSpec extends BaseJsonFormattersSpec with TableDrivenPropertyChe
       }
     }
 
+    "apply throws when given a bad value" in {
+      intercept[Exception] {
+        GrantLength.apply(5, "hours")
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (anjumabbas5) [reply src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala:51]: We already have one "read from json where grant length is Duration"

[the remark this answers is not in the document]
Comment (johnsgp) [reply src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala:51]: That one is not testing GrantLength using the Duration though. I was thinking of a test with:
```GrantLength.apply(1, "DAYS")```

[the remark this answers is not in the document]
Comment (anjumabbas5) [reply src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala:55]: In the test I mentioned above the grant length reader will call the 2-parameter `.apply`. Hence it showing as covered in the coverage report.

[the remark this answers is not in the document]
Comment (anjumabbas5) [reply src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala:61]: @johnsgp saying that I've added a test to GrantLengthSpec

File: src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala [hunk 4]
```diff
@@ -16,45 +16,116 @@
...
+  private val unitRead: Reads[String] = (JsPath \ "unit").read[String].map(u => u.toUpperCase)
+
+  implicit val durationReads: Reads[Duration] = (
+    amountRead and unitRead.map(ChronoUnit.valueOf(_))
+  )(Duration.of(_, _))
+
+  implicit val durationWrites: Writes[Duration] = (
+    (JsPath \ "amount").write[Long] and
+      (JsPath \ "unit").write[String]
+  )(d => (d.toSeconds, ChronoUnit.SECONDS.name()))
```
Comment (johnsgp) [inline src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala:48]: If you publish this library locally and test it with, say, third-party-application, what happens with the repository reads and writes? Is there a clash? Do they uses these formatters or db-specific ones?

File: src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala [hunk 5]
```diff
@@ -53,11 +55,16 @@ class GrantLengthSpec extends BaseJsonFormattersSpec with TableDrivenPropertyChe
...
-      Json.fromJson[GrantLength](JsNumber(2)) match {
-        case e: JsError if (JsError.Message.unapply(e) == Some("Invalid grant length 2")) => succeed
-        case _                                                                            => fail("Should have failed validation")
+    "read Int grant length into Period" in {
+      val grantLength = GrantLength.ONE_DAY
+      Json.fromJson[GrantLength](JsNumber(1)) shouldBe JsSuccess(grantLength)
+    }
+
+    "not read invalid json but throw an IllegalStateException instead" in {
+      val e = intercept[IllegalStateException] {
```
Comment (AndySpaven) [inline src/test/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLengthSpec.scala:64]: This should not have thrown but instead returned a JsError

File: src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala [hunk 6]
```diff
@@ -69,10 +108,11 @@ object GrantLength {
...
   import play.api.libs.json.Reads._
 
-  implicit val writesGrantLength: Writes[GrantLength] = implicitly[Writes[Int]].contramap(_.days)
+  implicit val writesGrantLength: Writes[GrantLength] = implicitly[Writes[Int]].contramap(x => x.period.getDays)
 
-  implicit val readsGrantLength: Reads[GrantLength] = implicitly[Reads[Int]].flatMapResult {
-    case i if (allowedIntegerValues.contains(i)) => JsSuccess(GrantLength.unsafeApply(i))
-    case i                                       => JsError(s"Invalid grant length $i")
+  implicit val readsGrantLength: Reads[GrantLength] = {
+    JsPath.read[Period].flatMapResult {
```
Comment (AndySpaven) [inline src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala:114]: This needs to not throw an exception - perhaps change this apply to also return option then pattern match or fold on the option of grantlength to return (js)success or (js)error

File: src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala [hunk 7]
```diff
@@ -16,45 +16,84 @@
...
+
+  private val errorMsg: String = "It should only be one of ('0 days, 1 day', '1 month', '3 months', '6 months', '1 year', '18 months', " +
+    "'3 years', '5 years', '10 years', '100 years')"
+
   def apply(grantLengthInDays: Int): Option[GrantLength] = {
-    GrantLength.values.find(e => e.days == grantLengthInDays)
+    GrantLength.values.find(e => e.period.getDays == grantLengthInDays)
+  }
+
+  def apply(period: Period): GrantLength = {
```
Comment (AndySpaven) [inline src/main/scala/uk/gov/hmrc/apiplatform/modules/applications/core/domain/models/GrantLength.scala:91]: Probably need to change to  option of grantlength (see comments below)

**Review by AndySpaven — CHANGES_REQUESTED** [review]

**Review by AndySpaven — APPROVED** [review]

**Pull request merged**